nomad/structs/node.go

```diff
@@ -1,11 +1,160 @@
+type CSITopology struct {
+	Segments map[string]string
+}
+
+func (t *CSITopology) Copy() *CSITopology {
+	if t == nil {
+		return nil
+	}
+
+	return &CSITopology{
+		Segments: helper.CopyMapStringString(t.Segments),
+	}
+}
+
+// CSINodeInfo is the fingerprinted data from a CSI Plugin that is specific to
+// the Node API.
+type CSINodeInfo struct {
+	// ID is the identity of a given nomad client as observed by the storage
+	// provider.
+	ID string
+
+	// MaxVolumes is the maximum number of volumes that can be attached to the
+	// current host via this provider.
+	// If 0 then unlimited volumes may be attached.
+	MaxVolumes int64
+
+	// AccessibleTopology specifies where (regions, zones, racks, etc.) the node is
+	// accessible from within the storage provider.
+	//
+	// A plugin that returns this field MUST also set the `RequiresTopologies`
+	// property.
+	//
+	// This field is OPTIONAL. If it is not specified, then we assume that the
+	// the node is not subject to any topological constraint, and MAY
+	// schedule workloads that reference any volume V, such that there are
+	// no topological constraints declared for V.
+	//
+	// Example 1:
+	//   accessible_topology =
+	//     {"region": "R1", "zone": "Z2"}
+	// Indicates the node exists within the "region" "R1" and the "zone"
+	// "Z2" within the storage provider.
+	AccessibleTopology *CSITopology
+}
+
+func (n *CSINodeInfo) Copy() *CSINodeInfo {
+	if n == nil {
+		return nil
+	}
+
+	nc := new(CSINodeInfo)
+	*nc = *n
+	nc.AccessibleTopology = n.AccessibleTopology.Copy()
+
+	return nc
+}
+
+// CSIControllerInfo is the fingerprinted data from a CSI Plugin that is specific to
+// the Controller API.
+type CSIControllerInfo struct {
+	Capabilities []string
+}
+
+// CSIInfo is the current state of a single CSI Plugin. This is updated regularly
+// as plugin health changes on the node.
+type CSIInfo struct {
+	PluginID          string
+	Healthy           bool
+	HealthDescription string
+	UpdateTime        time.Time
+
+	// RequiresControllerPlugin is set when the CSI Plugin returns the
+	// CONTROLLER_SERVICE capability. When this is true, the volumes should not be
+	// scheduled on this client until a matching controller plugin is available.
+	RequiresControllerPlugin bool
+
+	// RequiresTopologies is set when the CSI Plugin returns the
+	// VOLUME_ACCESSIBLE_CONSTRAINTS capability. When this is true, we must
+	// respect the Volume and Node Topology information.
+	RequiresTopologies bool
+
+	// CSI Specific metadata
+	ControllerInfo *CSIControllerInfo `json:",omitempty"`
+	NodeInfo       *CSINodeInfo       `json:",omitempty"`
+}
+
+func (c *CSIInfo) Copy() *CSIInfo {
+	if c == nil {
+		return nil
+	}
+
+	nc := new(CSIInfo)
+	*nc = *c
+	nc.NodeInfo = c.NodeInfo.Copy()
+
+	return nc
+}
+
+func (c *CSIInfo) SetHealthy(hs bool) {
+	c.Healthy = hs
+	if hs {
+		c.HealthDescription = "healthy"
+	} else {
+		c.HealthDescription = "unhealthy"
+	}
+}
+
+func (c *CSIInfo) IsEqual(o *CSIInfo) bool {
```

Comment:
I think we usually define the method as `Equals` and make it handle either pointer being nil.